File: Applications/Examples/perl/market_price_posting.pl
#|-----------------------------------------------------------------------------
#|            This source code is provided under the Apache 2.0 license      --
#|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
#|                See the project's LICENSE.md for details.                  --
#|           Copyright (C) Refinitiv 2019. All rights reserved.              --
#|-----------------------------------------------------------------------------


#!/usr/bin/perl
# Simple example of outputting Market Price JSON data using Websockets

use feature qw(say);
use JSON;
use Mojo::UserAgent;
use Getopt::Long;
use Time::HiRes;
use Socket;
use Sys::Hostname;
use utf8;
use Encode;

# Global Default Variables
$hostname = "127.0.0.1";
$port = "15000";
$user = "root";
$app_id = "256";
$position = inet_ntoa(scalar(gethostbyname(hostname())) || 'localhost');
$help = "";

# Global Variables
$web_socket_open = 0;
$next_post_time = 0;
$post_id = 1;

use sigtrap 'handler' => \&myhand, 'INT';

sub myhand
{
    kill 6, $$; # ABRT = 6
}

# Get command line parameters
GetOptions ('hostname=s' => \$hostname, 'port=s' => \$port, 'user=s' => \$user, 'app_id=s' => \$app_id, 'position=s' => \$position, 'help' => \$help);

if(not $help eq "") {
	print "Usage: market_price.pl [--hostname hostname] [--port port] [--app_id app_id] [--user user] [--position position] [--help]\n";
	exit 0;
}

# Generate a login request from command line data (or defaults) and send
sub send_login_request {
	my ($tx) = @_;
	my %login_json_hash = (
		'ID' => 1,
		'Domain' => 'Login',
		'Key' => {
			'Name' => '',
			'Elements' => {
					'ApplicationId' => '',
					'Position' => ''
			}
		}
	);

	$login_json_hash{'Key'}{'Name'} = $user;
	$login_json_hash{'Key'}{'Elements'}{'ApplicationId'} = $app_id;
	$login_json_hash{'Key'}{'Elements'}{'Position'} = $position;
	
	my $json = encode_json \%login_json_hash;
	my $json_pretty = JSON->new->pretty->encode(\%login_json_hash);
	$tx->send($json);
	print "SENT:\n";
	print "$json_pretty\n";
}

# Parse at high level and output JSON of message
sub process_message {
	my ($tx, $message_json) = @_;
	
	my $message_type = $message_json->{'Type'};
	
	if ($message_type eq "Refresh") {
		my $message_domain = $message_json->{'Domain'};
		if (not($message_domain eq "")) {
			if ($message_domain eq "Login") {
				send_market_price_request($tx);
			}
		}

		if ($message_json->{'ID'} == 2)
		{
			if ($next_post_time == 0 and 
					($message->{'State'} eq "" or $message->{'State'}{'Stream'} eq "Open" and $message->{'State'}{'Data'} eq "Ok")) {
				# Item stream is open. We can start posting.
				$next_post_time = Time::HiRes::time() + 3;
			}
		}
	} elsif ($message_type eq "Ping") {

		my %pong_json_hash = (
			'Type' => 'Pong'
		);

		my $json = encode_json \%pong_json_hash;
		my $json_pretty = JSON->new->pretty->encode(\%pong_json_hash);
		$tx->send($json);
		print "SENT:\n";
		print "$json_pretty\n";
	}
}

# Create and send simple Market Price request
sub send_market_price_request {
	my ($tx) = @_;
	my %mp_req_json_hash = (
        'ID' => 2,
        'Key' => {
            'Name' => 'TRI.N',
        },
	);

	my $json = encode_json \%mp_req_json_hash;
	my $json_pretty = JSON->new->pretty->encode(\%mp_req_json_hash);
	$tx->send($json);
	print "SENT:\n";
	print "$json_pretty\n";
}


# Create and send simple Market Price post
sub send_market_price_post {
	my ($tx) = @_;

    my %mp_post_json_hash = (
        'ID' =>  2,
        'Type' => 'Post',
        'Domain' => 'MarketPrice',
		'Ack' => JSON::true,
		'PostID' => $post_id,
        'PostUserInfo' =>  {
			# Use the IP address as the Post User Address
            'Address' => $position, 

			# Use the current process ID as the Post User Id
            'UserID' => int($$)
        },
		'Message' => {
			'ID' => 0,
			'Type' => 'Update',
			'Domain' => 'MarketPrice',
			'Fields' => {'BID' => 45.55,'BIDSIZE' => 18,'ASK' => 45.57,'ASKSIZE' => 19}
		}
    );

	my $json = encode_json \%mp_post_json_hash;
	my $json_pretty = JSON->new->pretty->encode(\%mp_post_json_hash);
	$tx->send($json);
	print "SENT:\n";
	print "$json_pretty\n";

	$post_id += 1
}

# Start websocket handshake
my $ua = Mojo::UserAgent->new;
my $ws_address = "ws://$hostname:$port/WebSocket";
print "Connecting to WebSocket $ws_address ...\n";
$ua->inactivity_timeout(0);
$ua->websocket($ws_address => ['tr_json2'] => sub {
	my ($ua, $tx) = @_;
	say 'WebSocket handshake failed!' and return unless $tx->is_websocket;
	say 'Subprotocol negotiation failed!' and return unless $tx->protocol;
	# Called when websocket is closed
	$tx->on(finish => sub {
		my ($tx, $code, $reason) = @_;
		say "WebSocket closed with status $code.";
	});
	# Called when message received, parse message into JSON for processing
	$tx->on(message => sub {
		my ($tx, $message) = @_;
		print "RECEIVED:\n";
		my @json_array = @{decode_json encode_utf8($message)};

		my $json_pretty = JSON->new->pretty->encode(\@json_array);
		print(encode_utf8($json_pretty));
		
		foreach my $single_message (@json_array) {
			process_message($tx, $single_message);
		}
	});
	
	print "WebSocket successfully connected!\n";
	$web_socket_open = 1;
	Mojo::IOLoop->recurring(0 => sub {
		if($web_socket_open and $next_post_time > 0 and Time::HiRes::time() >= $next_post_time) {
			send_market_price_post($tx);
			$next_post_time = Time::HiRes::time() + 3;
		}
	});
	send_login_request($tx);
});

# Event loop
Mojo::IOLoop->start unless Mojo::IOLoop->is_running;

File: Applications/Examples/R/market_price_posting.R
#|-----------------------------------------------------------------------------
#|            This source code is provided under the Apache 2.0 license 
#|  and is provided AS IS with no warranty or guarantee of fit for purpose.
#|                See the project's LICENSE.md for details.
#|           Copyright (C) 2017-2020,2024 LSEG. All rights reserved.
#|-----------------------------------------------------------------------------


# Simple example of outputting Market Price JSON data using Websockets

require("websockets") # https://github.com/brettjbush/R-Websockets
require("jsonlite")
require("GetoptLong")

config = ""
if(.Platform$OS.type == "unix") {
  config = system("ifconfig", intern=TRUE)
  ips = config[grep("inet addr", config)]
  if(length(ips) == 0) {
    ips = config[grep("inet ", config)]
    positions = strsplit(gsub(".*inet ([[:digit:]])", "\\1", ips), " ")
    first_positions = lapply(positions, '[[', 1)
    position = first_positions[first_positions != "127.0.0.1"][[1]]
  }
  else {
    position = strsplit(gsub(".*inet addr:([[:digit:]])", "\\1", ips), " ")[[1]][[1]]
  }
} else {
  config = system("ipconfig", intern=TRUE)
  ips = config[grep("IPv4", config)]
  position = strsplit(gsub(".*? ([[:digit:]])", "\\1", ips), " ")[[1]][[1]]
}

# Global Default Variables
hostname = "127.0.0.1"
port = "15000"
user = "root"
app_id = "256"

# Get command line parameters
GetoptLong(
  "hostname=s","",
  "port=s","",
  "user=s","",
  "app_id=s","",
  "position=s",""
)

next_post_time = 0
post_id = 1

# Start websocket handshake
ws_address = paste("ws://", hostname, ":", port, "/WebSocket", sep="")
cat(paste("Connecting to WebSocket", ws_address, "...\n"))
con = websocket(ws_address, port=as.integer(port), subprotocol="tr_json2", version=13)

# Create and send simple Market Price request
send_market_price_request = function(con) {
  mp_req_json_string = "{\"ID\":2,\"Key\":{\"Name\":\"TRI.N\"}}"
  mp_req_json = fromJSON(mp_req_json_string)
  websocket_write(mp_req_json_string, con)
  cat("SENT:\n")
  cat(toJSON(mp_req_json, pretty=TRUE, auto_unbox=TRUE))
  cat("\n")
}

# Create and send simple Market Price post
send_market_price_post = function(con) {
  mp_post_json_string =
		"{ \"ID\": 2, \"Type\": \"Post\", \"Domain\": \"MarketPrice\",
            \"Ack\": true, \"PostID\": <POSTID>,
			\"PostUserInfo\": { \"Address\": \"<ADDRESS>\", \"UserID\": <USERID> },
            \"Message\": {
                \"ID\":0,
                \"Type\":\"Update\",
                \"Fields\": {\"BID\": 45.55, \"BIDSIZE\": 18, \"ASK\": 45.57, \"ASKSIZE\": 19}
            }
		}"

	#Use the IP address as the Post User Address
	mp_post_json_string = gsub("<ADDRESS>", position, mp_post_json_string)

	#Use the current process ID as the Post User ID
	mp_post_json_string = gsub("<USERID>", Sys.getpid(), mp_post_json_string)

	mp_post_json_string = gsub("<POSTID>", post_id, mp_post_json_string)

  mp_post_json = fromJSON(mp_post_json_string)
  websocket_write(mp_post_json_string, con)
  cat("SENT:\n")
  cat(toJSON(mp_post_json, pretty=TRUE, auto_unbox=TRUE))
  cat("\n")

  next_post_time <<- Sys.time() + 3
  post_id <<- post_id + 1
}

# Parse at high level and output JSON of message
process_message = function(con, message_json) {
  message_type = message_json$Type

  if(message_type == "Refresh") {
	message_domain = message_json$Domain
	if(!is.null(message_domain)) {
	  if(message_domain == "Login") {
	    send_market_price_request(con)
	  }
	}

    if (as.integer(message_json$ID) == 2)
    {
        if (next_post_time == 0 &&
                (is.null(message_json$State) || message_json$State$Stream == "Open" && message_json$State$Data == "Ok"))
        {
            next_post_time <<- Sys.time() + 3
        }
    }
  } else if (message_type == "Ping") {
    pong_json_string = "{\"Type\":\"Pong\"}"
    pong_json = fromJSON(pong_json_string)
    websocket_write(pong_json_string, con)
    cat("SENT:\n")
    cat(toJSON(pong_json, pretty=TRUE, auto_unbox=TRUE))
    cat("\n")
  }
}

# Generate a login request from command line data (or defaults) and send
send_login_request = function(con) {
  login_json_string = "{\"ID\":1,\"Domain\":\"Login\",\"Key\":{\"Name\":\"<USER>\",\"Elements\":{\"ApplicationId\":\"<APP_ID>\",\"Position\":\"<POSITION>\"}}}"
  login_json_string = gsub("<USER>", user, login_json_string)
  login_json_string = gsub("<APP_ID>", app_id, login_json_string)
  login_json_string = gsub("<POSITION>", position, login_json_string)
  login_json = fromJSON(login_json_string)
  websocket_write(login_json_string, con)
  cat("SENT:\n")
  cat(toJSON(login_json, pretty=TRUE, auto_unbox=TRUE))
  cat("\n")
}

# Called when handshake is complete and websocket is open, send login
e = function(WS) {
  cat("WebSocket successfully connected!\n")
  send_login_request(con)
}
set_callback("established", e, con)

# Called when message received, parse message into JSON for processing
r = function(DATA,WS,...) {
  message = rawToChar(DATA)
  jsonArray = fromJSON(message, simplifyDataFrame=FALSE)

  cat("RECEIVED:\n")
  cat(toJSON(jsonArray, pretty=TRUE, auto_unbox=TRUE))
  cat("\n")

  for (singleMsg in jsonArray)
     process_message(con, singleMsg)
}
set_callback("receive", r, con)

# Called when websocket is closed
c = function(WS) {
  cat("WebSocket Closed\n")
}
set_callback("closed", c, con)

while (TRUE) {
  service(con)
  if (next_post_time != 0 && Sys.time() > next_post_time) {
    send_market_price_post(con)
  }
}
